edx_rest_framework_extensions/auth/jwt/decoder.py

```diff
@@ -173,9 +173,20 @@ def _set_filters(token):
 
 def _verify_jwt_signature(token, jwt_issuer, decode_symmetric_token):
     key_set = _get_signing_jwk_key_set(jwt_issuer, add_symmetric_keys=decode_symmetric_token)
-
     try:
-        _ = JWS().verify_compact(token, key_set)
+        for i in range(0, len(key_set)):
+            try:
+                _ = jwt.decode(
+                        token,
+                        key=key_set[i].key,
+                        algorithms=['HS256', 'RS512'],
```

Comment:
I think this is still vulnerable. One of these is a symmetric algorithm and the other is asymmetric. If you allow both, that means the attacker can take your RS512 _public_ key, then use it in HS256 as the _secret_ key, which will then allow verification to succeed. We need to ensure that the keys we supply here match the intended use of each key.

More generally, we should be relying on a JWT library to do this kind of looping and key-matching operation, rather than rewriting it ourselves, since it's a very delicate and error-prone part of JWT cryptography.

Reply:
I believe that in the standard JWT algorithm-confusion attack, the attacker needs to acquire the RSA public key and use it to sign via the HS256 algorithm. So you'd have to do something more custom on the token generation side in order to demonstrate this. More details: https://portswigger.net/web-security/jwt/algorithm-confusion

Oof, that code you linked to in pyjwkest sure doesn't inspire confidence. It looks like it trusts the `"alg"` field in the untrusted header, unless the JWK has an `"alg"` field of its own. :-/

Anyway, I think there's a not-too-terrible solution: Change `_get_signing_jwk_key_set` to return a list of (alg, key) tuples so that the algorithm always matches the key. (It could check if the key has an alg already defined, and if not, fall back to a hardcoded RS256 or HS256).  This is potentially a breaking change, since some deployers may not have an alg field defined in their keys, but I think for edx-platform we'll need to do a key rotation anyhow.

Reply:
(Ah, pyjwkest has code that narrows the keyset to just the keys with a matching algorithm -- see `alg2keytype` and `pick_keys`.)